src-self-hosted/link.zig

```diff
@@ -89,447 +115,570 @@ pub fn openBinFile(allocator: *Allocator, file: fs.File, options: Options) !ElfF
-
-    text_section_index: ?u16 = null,
-    symtab_section_index: ?u16 = null,
-    got_section_index: ?u16 = null,
-
-    /// The same order as in the file. ELF requires global symbols to all be after the
-    /// local symbols, they cannot be mixed. So we must buffer all the global symbols and
-    /// write them at the end. These are only the local symbols. The length of this array
-    /// is the value used for sh_info in the .symtab section.
-    local_symbols: std.ArrayListUnmanaged(elf.Elf64_Sym) = std.ArrayListUnmanaged(elf.Elf64_Sym){},
-    global_symbols: std.ArrayListUnmanaged(elf.Elf64_Sym) = std.ArrayListUnmanaged(elf.Elf64_Sym){},
-
-    local_symbol_free_list: std.ArrayListUnmanaged(u32) = std.ArrayListUnmanaged(u32){},
-    global_symbol_free_list: std.ArrayListUnmanaged(u32) = std.ArrayListUnmanaged(u32){},
-    offset_table_free_list: std.ArrayListUnmanaged(u32) = std.ArrayListUnmanaged(u32){},
-
-    /// Same order as in the file. The value is the absolute vaddr value.
-    /// If the vaddr of the executable program header changes, the entire
-    /// offset table needs to be rewritten.
-    offset_table: std.ArrayListUnmanaged(u64) = std.ArrayListUnmanaged(u64){},
-
-    phdr_table_dirty: bool = false,
-    shdr_table_dirty: bool = false,
-    shstrtab_dirty: bool = false,
-    offset_table_count_dirty: bool = false,
-
-    error_flags: ErrorFlags = ErrorFlags{},
-
-    /// A list of text blocks that have surplus capacity. This list can have false
-    /// positives, as functions grow and shrink over time, only sometimes being added
-    /// or removed from the freelist.
-    ///
-    /// A text block has surplus capacity when its overcapacity value is greater than
-    /// minimum_text_block_size * alloc_num / alloc_den. That is, when it has so
-    /// much extra capacity, that we could fit a small new symbol in it, itself with
-    /// ideal_capacity or more.
-    ///
-    /// Ideal capacity is defined by size * alloc_num / alloc_den.
-    ///
-    /// Overcapacity is measured by actual_capacity - ideal_capacity. Note that
-    /// overcapacity can be negative. A simple way to have negative overcapacity is to
-    /// allocate a fresh text block, which will have ideal capacity, and then grow it
-    /// by 1 byte. It will then have -1 overcapacity.
-    text_block_free_list: std.ArrayListUnmanaged(*TextBlock) = std.ArrayListUnmanaged(*TextBlock){},
-    last_text_block: ?*TextBlock = null,
-
-    /// `alloc_num / alloc_den` is the factor of padding when allocating.
-    const alloc_num = 4;
-    const alloc_den = 3;
-
-    /// In order for a slice of bytes to be considered eligible to keep metadata pointing at
-    /// it as a possible place to put new symbols, it must have enough room for this many bytes
-    /// (plus extra for reserved capacity).
-    const minimum_text_block_size = 64;
-    const min_text_capacity = minimum_text_block_size * alloc_num / alloc_den;
+pub const File = struct {
```

Comment:
The way you split this up into `File.Elf` and `File.C` is :+1:

Reply:
:)